### PR TITLE
feat: transform sequelize data types to json schema

### PR DESCRIPTION
**Pull request by ayeo127**

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module.exports = {
+  env: {
+    browser: true,
```
Comment (RickEyre) [inline .eslintrc.js]: How come this is set? Isn't this purely Node?

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  plugins: ['jest'],
+  globals: {
+    Atomics: 'readonly',
+    SharedArrayBuffer: 'readonly'
+  },
+  parserOptions: {
+    ecmaVersion: 2018,
+    sourceType: 'module'
+  },
+  rules: {
```
Comment (RickEyre) [inline .eslintrc.js]: Can we remove this?

File: .babelrc [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "presets": ["@babel/preset-env"]
+}
```
Comment (RickEyre) [inline .babelrc]: Add newline please.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "@lemonpeach/sequelize-to-json-schema",
+  "version": "0.0.0",
+  "description": "",
+  "main": "lib/index.js",
+  "scripts": {
+    "clean": "shx rm -rf lib",
+    "build": "npm run clean && babel src -d lib",
+    "lint": "eslint src",
+    "test": "jest src"
```
Comment (RickEyre) [inline package.json]: Can we add a lint step into this before jest?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Lemonpeach/sequelize-to-json-schema.git"
+  },
+  "author": "",
+  "license": "ISC",
```
Comment (RickEyre) [inline package.json]: I believe this is the wrong license.  It should be MPL2. I forget the version code.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    "clean": "shx rm -rf lib",
+    "build": "npm run clean && babel src -d lib",
+    "lint": "eslint src",
+    "test": "jest src"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Lemonpeach/sequelize-to-json-schema.git"
+  },
+  "author": "",
```
Comment (RickEyre) [inline package.json]: Remove this, we can add an `AUTHORS` file after.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "@lemonpeach/sequelize-to-json-schema",
+  "version": "0.0.0",
+  "description": "",
```
Comment (RickEyre) [inline package.json]: Can you add a description please.

File: package.json [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    "type": "git",
+    "url": "git+https://github.com/Lemonpeach/sequelize-to-json-schema.git"
+  },
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/Lemonpeach/sequelize-to-json-schema/issues"
+  },
+  "homepage": "https://github.com/Lemonpeach/sequelize-to-json-schema#readme",
+  "dependencies": {},
```
Comment (RickEyre) [inline package.json]: Remove this please.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    "eslint-config-standard": "^14.1.0",
+    "eslint-plugin-import": "^2.18.2",
+    "eslint-plugin-jest": "^22.16.0",
+    "eslint-plugin-node": "^9.2.0",
+    "eslint-plugin-promise": "^4.2.1",
+    "eslint-plugin-standard": "^4.0.1",
+    "jest": "^24.9.0",
+    "lodash": "^4.17.15",
+    "pg": "^7.12.1",
+    "sequelize": "^5.18.1",
```
Comment (RickEyre) [inline package.json:35]: This needs to be a `peerDep`

File: src/__tests__/json-schema.test.js [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+      default: () => '0ef9f424-c6f0-4bfe-b55e-87c50420180f'
+    }))
+  })
+
+  afterEach(() => {
+    jest.resetModules()
+    jest.clearAllMocks()
+  })
+
+  test.each(SEQUELIZE_JSON_TABLE)(
```
Comment (RickEyre) [inline src/__tests__/json-schema.test.js:174]: Woah, I had no idea this was a thing. Nice!

File: src/__tests__/json-schema.test.js [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    'should handle type %s',
+    (type, _sequelize, json) => {
+      const { jsonSchema } = require('../json-schema')
+      const model = sequelize.define(type, {
+        id: {
+          type: _sequelize,
+          primaryKey: true
```
Comment (RickEyre) [inline src/__tests__/json-schema.test.js]: How come there is a primary key on each of these?

File: src/__tests__/json-schema.test.js [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+
+  test('should handle default values', () => {
+    const Sequelize = require('sequelize')
+    const { jsonSchema } = require('../json-schema')
+
+    const defaultUuidModel = sequelize.define('default-uuid', {
+      id: {
+        type: Sequelize.UUID,
+        defaultValue: Sequelize.UUIDV4,
+        primaryKey: true
```
Comment (RickEyre) [inline src/__tests__/json-schema.test.js]: Same question about `primaryKey` here too.

File: src/__tests__/json-schema.test.js [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        },
+        updatedAt: {
+          type: 'date',
+          default: undefined
+        }
+      },
+      required: ['id', 'createdAt', 'updatedAt']
+    })
+
+    const notRequiredModel = sequelize.define('not-required', {
```
Comment (RickEyre) [inline src/__tests__/json-schema.test.js]: Why this case? What's different about it?

**Review by RickEyre — CHANGES_REQUESTED** [review]

File: .eslintrc.js (hunk 1 above)
Comment (ayeo127) [reply .eslintrc.js]: yea, it got produced when I did eslint init. Removed

File: .eslintrc.js (hunk 2 above)
Comment (ayeo127) [reply .eslintrc.js]: removed

File: .babelrc (hunk 3 above)
Comment (ayeo127) [reply .babelrc]: done

File: package.json (hunk 4 above)
Comment (ayeo127) [reply package.json]: done

File: package.json (hunk 5 above)
Comment (ayeo127) [reply package.json]: MPL2.0 done

File: package.json (hunk 9 above)
Comment (ayeo127) [reply package.json:35]: done

File: src/__tests__/json-schema.test.js (hunk 11 above)
Comment (ayeo127) [reply src/__tests__/json-schema.test.js]: removed

File: src/__tests__/json-schema.test.js (hunk 12 above)
Comment (ayeo127) [reply src/__tests__/json-schema.test.js]: removed

File: src/__tests__/json-schema.test.js (hunk 13 above)
Comment (ayeo127) [reply src/__tests__/json-schema.test.js]: removed

File: src/__tests__/json-schema.test.js (hunk 12 above)
Comment (RickEyre) [reply src/__tests__/json-schema.test.js]: Looks like this one is still here?

File: src/__tests__/json-schema.test.js [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+        }
+      },
+      required: ['createdAt', 'updatedAt']
+    })
+
+    const defaultStringModel = sequelize.define('default-string', {
+      id: {
+        type: Sequelize.STRING,
+        defaultValue: 'wonkey watermelon',
+        primaryKey: true
```
Comment (RickEyre) [inline src/__tests__/json-schema.test.js]: And here.

File: src/__tests__/json-schema.test.js [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+
+  test('should handle required values', () => {
+    const Sequelize = require('sequelize')
+    const { jsonSchema } = require('../json-schema')
+
+    const requiredModel = sequelize.define('required-uuid', {
+      id: {
+        type: Sequelize.UUID,
+        defaultValue: Sequelize.UUIDV4,
+        primaryKey: true,
```
Comment (RickEyre) [inline src/__tests__/json-schema.test.js]: And here.

**Review by RickEyre — CHANGES_REQUESTED** [review]

File: src/__tests__/json-schema.test.js [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+      'string',
+      Sequelize.CHAR,
+      { type: 'string' }
+    ],
+    [
+      'string 123',
+      Sequelize.STRING(1234),
+      {
+        type: 'string',
+        maxLength: 1234
```
Comment (RickEyre) [inline src/__tests__/json-schema.test.js:20]: Does `maxLength` refer to character length or byte length in JSON schema? And similarly does `Sequelize.STRING(1234)` refer to max character length or byte length? We shouldn't conflate the two.

File: src/__tests__/json-schema.test.js [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+      'float',
+      Sequelize.FLOAT,
+      { type: 'number' }
+    ],
+    [
+      'float 11',
+      Sequelize.FLOAT(11),
+      {
+        type: 'number',
+        maximum: 11
```
Comment (RickEyre) [inline src/__tests__/json-schema.test.js]: Same question here as for string length.

**Review by RickEyre — CHANGES_REQUESTED** [review]

File: src/__tests__/json-schema.test.js (hunk 16 above)
Comment (ayeo127) [reply src/__tests__/json-schema.test.js:20]: json's maxLength refers to max char length.
sequelize's length refers to char length. There is a boolean option for binary, but it doesnt seem to affect the length when its binary=true https://github.com/sequelize/sequelize/blob/master/lib/data-types.js#L62.

File: src/__tests__/json-schema.test.js (hunk 17 above)
Comment (ayeo127) [reply src/__tests__/json-schema.test.js]: json's maximum is the inclusive maximum number.

sequelize's floating type (Float, Real, Double...etc) length refers to the number of digits excluding the decimal. E.g. DOUBLE(8, 3) where 8 is the total no. of digits excluding the decimal point, and 3 is the no. of digits to follow the decimal. https://code.rohitink.com/2013/06/12/mysql-integer-float-decimal-data-types-differences/

**Review by RickEyre — APPROVED** [review]

**Pull request merged**